File: card/easy/easystrings/006.go
package easystrings

// 字符串转整数（atoi）
// https://leetcode-cn.com/explore/interview/card/top-interview-questions-easy/5/strings/37/
// https://leetcode-cn.com/problems/string-to-integer-atoi/description/

func myAtoi(str string) int {
	min := -(1 << 31)
	max := (1 << 31) - 1
	ans := 0
	sign := 1
	idx := 0
	length := len(str)
	for idx < length {
		if str[idx] == ' ' {
			idx++
		} else {
			break
		}
	}
	if idx < length {
		if str[idx] == '-' {
			idx++
			sign = -1
		} else if str[idx] == '+' {
			idx++
		}
	}
	for idx < length {
		ch := str[idx]
		if '0' <= ch && ch <= '9' {
			ans = ans*10 + int(ch-'0')*sign
			if ans < min {
				ans = min
				break
			} else if ans > max {
				ans = max
				break
			}
		} else {
			break
		}
		idx++
	}

	return ans
}

File: card/easy/linkedlist/006/006.cpp
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

// Linked List Cycle
// https://leetcode.com/explore/interview/card/top-interview-questions-easy/93/linked-list/773/
// https://leetcode.com/problems/linked-list-cycle/description/

struct ListNode {
    int val;
    struct ListNode *next;
};
// bool hasCycle(struct ListNode *head) {
    
//     if (head == NULL) {
//         return false;
//     }
    
//     struct ListNode *node = head->next;
//     struct ListNode *prev = head;
//     struct ListNode *next = NULL;
//     while(node != NULL && node != head){
//         next = node->next;
//         node->next = prev;
//         prev = node;
//         node = next;
//     }
    
//     return node == head;
// }

bool hasCycle(struct ListNode *head) {

    if (head == NULL || head->next == NULL) {
        return false;
    }

    struct ListNode *slow = head;
    struct ListNode *fast = head->next;
    
    while(slow != fast ){
        
        if (fast == NULL || fast->next == NULL) {
            return false;
        }
        slow = slow->next;
        fast = fast->next->next;
    }    
        
    return true;
}

void test1() {
    struct ListNode n1 = {1, NULL};
    struct ListNode n2 = {2, &n1};
    struct ListNode n3 = {3, &n2};
    struct ListNode n4 = {4, &n3};
    n1.next = &n4;

    assert(hasCycle(&n4));
}

void test2() {
    struct ListNode n4 = {-4, NULL};
    struct ListNode n3 = {0, &n4};
    struct ListNode n2 = {2, &n3};
    struct ListNode n1 = {3, &n2};
    n4.next = &n2;

    assert(hasCycle(&n4));

}

int main(int argc, char const *argv[])
{
    test1();
    test2();

    return 0;
}
